pkg/security/resolvers/tags/resolver_linux.go

```diff
@@ -117,30 +131,108 @@ func (t *LinuxResolver) checkTags(pendingWorkload *Workload) {
+	workload.Tags = newTags
+
+	// For container workloads, try to extract image information
+	if len(workload.ContainerID) != 0 {
+		workload.Selector.Image = utils.GetTagValue("image_name", newTags)
+		workload.Selector.Tag = utils.GetTagValue("image_tag", newTags)
+		if len(workload.Selector.Image) != 0 && len(workload.Selector.Tag) == 0 {
+			workload.Selector.Tag = "latest"
+		}
+	} else {
+		// For cgroup workloads, set service information as the selector
+		serviceName := utils.GetTagValue("service", newTags)
+		if len(serviceName) != 0 {
+			workload.Selector.Image = serviceName
+			workload.Selector.Tag = utils.GetTagValue("version", newTags)
+			if len(workload.Selector.Tag) == 0 {
+				workload.Selector.Tag = "latest"
+			}
+		}
 	}
 
 	return nil
 }
 
 // NewResolver returns a new tags resolver
-func NewResolver(tagger Tagger, cgroupsResolver *cgroup.Resolver) *LinuxResolver {
+func NewResolver(tagger Tagger, cgroupsResolver *cgroup.Resolver, versionResolver func(servicePath string) string) *LinuxResolver {
 	resolver := &LinuxResolver{
 		Notifier:             utils.NewNotifier[Event, *Workload](),
 		DefaultResolver:      NewDefaultResolver(tagger),
 		workloadsWithoutTags: make(chan *Workload, 100),
 		cgroupResolver:       cgroupsResolver,
+		versionResolver:      versionResolver,
 		workloads:            make(map[containerutils.CGroupID]*Workload),
 	}
 	return resolver
 }
+
+// ResolveWithErr overrides the default implementation to use Linux-specific workload resolution
+func (t *LinuxResolver) ResolveWithErr(id interface{}) ([]string, error) {
+	return t.resolveWorkloadTags(id)
+}
+
+// resolveWorkloadTags overrides the default implementation to handle CGroup resolution on Linux
+func (t *LinuxResolver) resolveWorkloadTags(id interface{}) ([]string, error) {
+	if id == nil {
+		return nil, fmt.Errorf("nil workload id")
+	}
+
+	switch v := id.(type) {
+	case containerutils.ContainerID:
+		if len(v) == 0 {
+			return nil, fmt.Errorf("empty container id")
+		}
+		// Resolve as a container ID
+		return GetTagsOfContainer(t.tagger, v)
+	case containerutils.CGroupID:
+		if len(v) == 0 {
+			return nil, fmt.Errorf("empty cgroup id")
+		}
+		// Generate systemd service tags for cgroup workloads
+		tags := t.getCGroupTags(v)
+		return tags, nil
+	default:
+		return nil, fmt.Errorf("unknown workload id type: %T", id)
+	}
+}
+
+// getCGroupTags generates tags for cgroup workloads (systemd services) with version resolution
+func (t *LinuxResolver) getCGroupTags(cgroupID containerutils.CGroupID) []string {
+	if len(cgroupID) == 0 {
+		return nil
+	}
+
+	systemdService := filepath.Base(string(cgroupID))
+	serviceVersion := ""
+	servicePath := filepath.Join(systemdSystemDir, systemdService)
+
+	// Try to resolve version using version resolver
+	if t.versionResolver != nil {
+		serviceVersion = t.versionResolver(servicePath)
+	}
+
+	tags := []string{
+		"service:" + systemdService,
+		"version:" + serviceVersion,
```

Comment:
I think it would be better not to add a `version` tag if we failed to get the service version